src/test/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImplTest.java

```diff
@@ -0,0 +1,160 @@
+import com.atlassian.bitbucket.jenkins.internal.credentials.JenkinsToBitbucketCredentials;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketCDCapabilities;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeployment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeploymentEnvironment;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.DeploymentState;
+import hudson.model.FreeStyleProject;
+import hudson.model.Item;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Answers;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.io.PrintStream;
+
+import static java.lang.String.format;
+import static java.util.Collections.emptySet;
+import static java.util.Collections.singleton;
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+import static org.mockito.Mockito.*;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DeploymentPosterImplTest {
+
+    private static final String BASE_URL = "http://localhost:7990/bitbucket";
+    private static final BitbucketDeploymentEnvironment ENVIRONMENT = new BitbucketDeploymentEnvironment("ENV-KEY", "Env name");
+    private static final BitbucketDeployment DEPLOYMENT = new BitbucketDeployment(1, "desc", "name",
+            ENVIRONMENT, "key", DeploymentState.FAILED, "url");
+    private static final String PROJECT_KEY = "projectKey";
+    private static final String REPO_SLUG = "repoSlug";
+    private static final String REVISION_SHA = "revisionSha";
+    private static final String SERVER_ID = "bbsServerId";
+    private static final String SERVER_NAME = "localhost";
+
+    @Mock
+    public BitbucketClientFactoryProvider clientFactoryProvider;
+
+    @InjectMocks
+    public DeploymentPosterImpl poster;
+    @Mock
+    public Run<FreeStyleProject, ?> run;
+    @Mock
+    public TaskListener taskListener;
+    @Mock
+    private BitbucketCredentials bitbucketCredentials;
+    @Mock(answer = Answers.RETURNS_DEEP_STUBS)
+    private BitbucketClientFactory clientFactory;
+    @Mock
+    private BitbucketTokenCredentials globalAdminCredentials;
+    @Mock
+    private GlobalCredentialsProvider globalCredentialsProvider;
+    @Mock
+    private JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials;
+    @Mock
+    private BitbucketPluginConfiguration pluginConfiguration;
+    @Mock
+    private PrintStream printStream;
+    @Mock
+    private BitbucketServerConfiguration server;
+
+    @Before
+    public void setup() {
+        FreeStyleProject parent = mock(FreeStyleProject.class);
+        when(run.getParent()).thenReturn(parent);
+        when(globalCredentialsProvider.getGlobalAdminCredentials()).thenReturn(of(globalAdminCredentials));
```

Comment:
I think it would also be good to have a test for null credentials (or anonymous_credentials if we're going down that path)